File: providers/adapters/llamaindex-bridge.ts
/**
 * LlamaIndex CodeSplitter Python Bridge
 *
 * TypeScript wrapper for calling LlamaIndex CodeSplitter via subprocess.
 * Uses tree-sitter parsing for semantic code chunking.
 *
 * Requirements:
 *   - Python 3.10+
 *   - LlamaIndex package: pip install llama-index-core
 */

import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { callPythonBridge, isPythonPackageAvailable } from "./python-bridge-utils.ts";

// Resolve path to the Python script
const __dirname = dirname(fileURLToPath(import.meta.url));
const SCRIPT_PATH = join(__dirname, "llamaindex_bridge.py");

/**
 * Chunk result from LlamaIndex.
 */
export interface LlamaIndexChunkResult {
	id: string;
	text: string;
	startLine: number;
	endLine: number;
}

/**
 * Options for LlamaIndex chunking.
 */
export interface LlamaIndexOptions {
	/** Maximum chunk size in characters */
	chunkSize: number;
	/** Python executable path (default: "python3") */
	pythonPath?: string;
}

/**
 * Call LlamaIndex CodeSplitter via subprocess.
 *
 * @param filepath - File path (used for language detection)
 * @param code - Source code content
 * @param options - Chunking options
 * @returns Array of chunk results
 */
export async function callLlamaIndex(
	filepath: string,
	code: string,
	options: LlamaIndexOptions,
): Promise<LlamaIndexChunkResult[]> {
	const pythonPath =
		options.pythonPath || process.env.LLAMAINDEX_PYTHON_PATH || "python3";
	const args = [SCRIPT_PATH, filepath, String(options.chunkSize)];

	try {
		return await callPythonBridge<LlamaIndexChunkResult[]>(
			SCRIPT_PATH,
			args,
			code,
			pythonPath,
			"LlamaIndex",
		);
	} catch (error) {
		throw new Error(
			`${error}\nMake sure Python is installed and LlamaIndex is available:\n` +
				`  pip install llama-index-core`,
		);
	}
}

/**
 * Check if LlamaIndex is available.
 */
export async function isLlamaIndexAvailable(
	pythonPath = "python3",
): Promise<boolean> {
	return isPythonPackageAvailable(
		pythonPath,
		"from llama_index.core.node_parser import CodeSplitter",
	);
}

File: providers/adapters/llamaindex_bridge.py
#!/usr/bin/env python3
"""
LlamaIndex CodeSplitter Python Bridge for superbench-bench-code-chunk.

This script provides a subprocess interface to LlamaIndex's CodeSplitter.
Uses tree-sitter parsing for semantic code chunking.

Usage:
    python llamaindex_bridge.py <filepath> <chunk_size>

    filepath: Path to the file (used for language detection)
    chunk_size: Maximum chunk size in characters

Code is read from stdin. Output is JSON array of chunks.

Install dependencies:
    pip install llama-index-core
    # Tree-sitter parsers are included with llama-index
"""

import json
import sys
from typing import Any


def get_language(filepath: str) -> str:
    """Determine programming language from file extension.

    Returns language name as expected by LlamaIndex CodeSplitter.
    """
    ext = filepath.rsplit(".", 1)[-1].lower() if "." in filepath else ""

    # Map extensions to LlamaIndex supported languages
    # See: https://docs.llamaindex.ai/en/stable/api_reference/node_parsers/code/
    lang_map = {
        "py": "python",
        "js": "javascript",
        "ts": "typescript",
        "tsx": "tsx",
        "jsx": "javascript",
        "rs": "rust",
        "go": "go",
        "java": "java",
        "c": "c",
        "cpp": "cpp",
        "h": "c",
        "hpp": "cpp",
        "rb": "ruby",
        "php": "php",
        "cs": "c_sharp",
        "swift": "swift",
        "kt": "kotlin",
        "scala": "scala",
        "lua": "lua",
        "pl": "perl",
        "r": "r",
        "sh": "bash",
        "bash": "bash",
        "zsh": "bash",
        "ps1": "powershell",
        "sql": "sql",
        "md": "markdown",
        "html": "html",
        "css": "css",
        "json": "json",
        "yaml": "yaml",
        "yml": "yaml",
        "toml": "toml",
        "xml": "xml",
    }
    return lang_map.get(ext, "python")  # Default to Python


def chunk_with_llamaindex(code: str, filepath: str, chunk_size: int) -> list[dict[str, Any]]:
    """Chunk using LlamaIndex CodeSplitter."""
    from llama_index.core import Document
    from llama_index.core.node_parser import CodeSplitter

    language = get_language(filepath)

    splitter = CodeSplitter(
        language=language,
        max_chars=chunk_size,
    )

    doc = Document(text=code)
    nodes = splitter.get_nodes_from_documents([doc])

    results = []
    for node in nodes:
        # LlamaIndex provides start_char_idx in metadata
        start_idx = node.start_char_idx if node.start_char_idx is not None else 0
        end_idx = node.end_char_idx if node.end_char_idx is not None else len(node.text)

        # Convert character offsets to line numbers (1-indexed)
        start_line = code[:start_idx].count("\n") + 1
        end_line = code[:end_idx].count("\n") + 1

        results.append({
            "id": f"{filepath}:{start_line}-{end_line}",
            "text": node.text,
            "startLine": start_line,
            "endLine": end_line,
        })

    return results


def main():
    if len(sys.argv) < 3:
        print(
            "Usage: llamaindex_bridge.py <filepath> <chunk_size>",
            file=sys.stderr,
        )
        print("  Code is read from stdin", file=sys.stderr)
        sys.exit(1)

    filepath = sys.argv[1]
    chunk_size = int(sys.argv[2]) if len(sys.argv) > 2 else 1500

    # Read code from stdin
    code = sys.stdin.read()

    try:
        results = chunk_with_llamaindex(code, filepath, chunk_size)
        print(json.dumps(results))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)


if __name__ == "__main__":
    main()
